Update FilePrefetchBuffer::Read to use FS buffer

archang19 opened the pull request.

archang19 commented on file/file_prefetch_buffer.cc:
```diff
@@ -530,14 +587,15 @@ Status FilePrefetchBuffer::PrefetchInternal(const IOOptions& opts,
                                             size_t readahead_size,
-                                            bool& copy_to_overlap_buffer) {
+                                            bool& copy_to_overlap_buffer,
+                                            bool& use_staging_buffer) {
```
I considered that but wanted to get your take (see "Discussion Points" in the PR description). I think I am currently leaning towards unifying the two buffers

archang19 replied:
(PR has been updated to only use `overlap_buf_`)

archang19 commented on file/file_prefetch_buffer.h:
```diff
@@ -465,6 +491,51 @@ class FilePrefetchBuffer {
+  // Whether we reuse the file system provided buffer
+  // Until we also handle the async read case, only enable this optimization
+  // for the synchronous case when num_buffers_ = 1.
+  bool UseFSBuffer(RandomAccessFileReader* reader) {
```
Agreed. Are you thinking of just a `use_fs_buffer` parameter, or can we pass in the entire `RandomAccessFileReader*` so that we don't need to pass it into every single one of the other methods?

archang19 commented on file/file_prefetch_buffer.h:
```diff
@@ -388,13 +405,21 @@ class FilePrefetchBuffer {
   void PrepareBufferForRead(BufferInfo* buf, size_t alignment, uint64_t offset,
                             size_t roundup_len, bool refit_tail,
-                            uint64_t& aligned_useful_len);
+                            uint64_t& aligned_useful_len, bool use_fs_buffer);
```
I updated all of the methods I updated and all the new methods I defined

archang19 commented on file/file_prefetch_buffer.h:
```diff
@@ -491,7 +566,7 @@ class FilePrefetchBuffer {
                            size_t readahead_size, uint64_t& offset,
                            uint64_t& end_offset, size_t& read_len,
-                           uint64_t& aligned_useful_len);
+                           uint64_t& aligned_useful_len, bool use_fs_buffer);
```
I updated all of the methods I updated and all the new methods I defined

archang19 commented on file/file_prefetch_buffer.h:
```diff
@@ -465,6 +491,51 @@ class FilePrefetchBuffer {
+      return s;
+    }
+    buf->buffer_.SetBuffer(
+        reinterpret_cast<char*>(read_req.fs_scratch.release()),
```
I updated my PR and I _think_ I it is what we want

anand1976 commented on file/file_prefetch_buffer.cc:
```diff
@@ -587,7 +650,7 @@ Status FilePrefetchBuffer::PrefetchInternal(const IOOptions& opts,
       // buffer and remaining will be updated below.
       size_t initial_buf_size = overlap_buf_->CurrentSize();
-      CopyDataToBuffer(buf, offset, length);
+      CopyDataToOverlapBuffer(buf, offset, length);
```
Should this entire block (i.e starting at line 648) be skipped if `use_fs_buffer` is true? Right now, its a bit hard to reason about that case. We may have already copied to overlap buffer, and we'll attempt to copy again here, which is confusing.

archang19 replied:
I agree that the logic is a bit hard to reason about. I have added an explanation as a comment (starting with `Note: why do we not end up performing a duplicate copy`. 

When `use_fs_buffer` is true and we have only 1 buffer, `buf->DoesBufferContainData()` will return false so we'll skip this block. When we have more than 1 buffer, we will have relevant data to be copied to the overlap buffer, since we always clean up "outdated" buffers.

archang19 replied:
To prove this is correct, I updated this PR to also have a `FilePrefetchBuffer::CopyDataToOverlapBuffer:Complete` sync point which I use in the tests to count the number of times we copy data to the overlap buffer. This helped me confirm that the overlap buffer copying logic worked as I had expected it to.

anand1976 commented on file/prefetch_test.cc:
```diff
@@ -3290,6 +3290,276 @@ TEST_F(FilePrefetchBufferTest, SyncReadaheadStats) {
+  std::string Path(const std::string& fname) { return test_dir_ + "/" + fname; }
+};
+
+TEST_F(FSBufferPrefetchTest, FSBufferPrefetchStatsInternals) {
```
Can we do these tests for both sync and async by parameterizing the test? Just want to ensure nothing is broken in the async case.

anand1976 reviewed (commented):
LGTM overall. A few minor comments.

archang19 commented on file/file_prefetch_buffer.cc:
```diff
@@ -194,10 +219,8 @@ void FilePrefetchBuffer::CopyDataToBuffer(BufferInfo* src, uint64_t& offset,
-         src->buffer_.BufferStart() + copy_offset, copy_len);
-
-  dst->buffer_.Size(dst->CurrentSize() + copy_len);
+  assert(copy_len <= dst->buffer_.Capacity() - dst->buffer_.CurrentSize());
```
`Append` already handles the overflow check in `aligned_buffer.h`, so adding that check there would require modifying its existing behavior

```cpp
size_t Append(const char* src, size_t append_size) {
    size_t buffer_remaining = capacity_ - cursize_;
    size_t to_copy = std::min(append_size, buffer_remaining);

    if (to_copy > 0) {
      memcpy(bufstart_ + cursize_, src, to_copy);
      cursize_ += to_copy;
    }
    return to_copy;
  }
```

archang19 commented on file/file_prefetch_buffer.h:
```diff
@@ -379,12 +389,21 @@ class FilePrefetchBuffer {
 
   void TEST_GetBufferOffsetandSize(
-      std::vector<std::pair<uint64_t, size_t>>& buffer_info) {
+      std::vector<std::tuple<uint64_t, size_t, bool>>& buffer_info) {
```
I added this third field to validate async prefetching, since otherwise I cannot distinguish between `async_req_len` and `CurrentSize()`

archang19 commented:
@anand1976 here is a summary of the changes since the last review, since the old comment thread seems to be hidden:

1. https://github.com/facebook/rocksdb/pull/13118#discussion_r1847234840 why we can't refactor our `assert` to `Append`
2.  https://github.com/facebook/rocksdb/pull/13118#discussion_r1849109263 why we do not end up "double copying" to the overlap buffer + tests that validate this using sync points
3. https://github.com/facebook/rocksdb/pull/13118#discussion_r1849082246 parametrized tests for both sync/async cases, with discussion on why the async testing is a bit tricky/annoying currently

anand1976 reviewed (approved):
LGTM. Can you add an entry under `unreleased_history/performance_improvements` so release history gets update?
